discovery/discovery.go

```diff
@@ -127,51 +127,48 @@ func (o *orchestratorPool) GetOrchestrators(ctx context.Context, numOrchestrator
 	}
 
 	var infos []*net.OrchestratorInfo
-	var suspendedInfos *suspensionQueue
-	var timedOut bool
-	var nbResp int
+	suspendedInfos := newSuspensionQueue()
+	timedOut := false
+	nbResp := 0
+	infoCh := make(chan *net.OrchestratorInfo, numAvailableOrchs)
+	errCh := make(chan error, numAvailableOrchs)
 
-	// try to discover orchestrators until at least 1 is found (with the exponential backoff timout)
-	timeout := getOrchestratorsCutoffTimeout
-	for {
-		ctx, cancel := context.WithTimeout(clog.Clone(context.Background(), ctx), timeout)
-		infoCh := make(chan *net.OrchestratorInfo, numAvailableOrchs)
-		errCh := make(chan error, numAvailableOrchs)
-
-		for _, uri := range uris {
-			go getOrchInfo(ctx, uri, infoCh, errCh)
-		}
+	ctx, cancel := context.WithTimeout(clog.Clone(context.Background(), ctx), maxGetOrchestratorCutoffTimeout)
+	for _, uri := range uris {
+		go getOrchInfo(ctx, uri, infoCh, errCh)
+	}
 
-		timedOut = false
-		suspendedInfos = newSuspensionQueue()
-		nbResp = 0
-		for i := 0; i < numAvailableOrchs && len(infos) < numOrchestrators && !timedOut; i++ {
-			select {
-			case info := <-infoCh:
-				if penalty := suspender.Suspended(info.Transcoder); penalty == 0 {
-					infos = append(infos, info)
-				} else {
-					heap.Push(suspendedInfos, &suspension{info, penalty})
-				}
-				nbResp++
-			case <-errCh:
-				nbResp++
-			case <-ctx.Done():
+	// try to wait for orchestrators until at least 1 is found (with the exponential backoff timout)
+	timeout := getOrchestratorsCutoffTimeout
+	timer := time.NewTimer(timeout)
```

Comment:
Do timers need to be cleaned up at all i.e. with Stop() if there is a chance that the `timer.C` channel is not drained before moving on? I'm thinking of the case where right now in the `<-timer.C` case we extend the timer no matter what and then move on to the next iteration of the loop, but if `timedOut` is set to true we exit at the next iteration. At this point, the timer was extended and its `timer.C` channel was not drained. Not sure if that is a problem? If it is, perhaps calling Stop() when we exit the loop or not extending the timer if we're going to exit the loop on the next iteration would work?

Reply:
Good question. I think it's ok to not drain (effectively close) the timer channel. At least according to [this thread](https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open), this is common and the channel will just be garbage collected.